Package 'rcutils' exports the library 'rcutils' which couldn't be found

### Generated by Generative AI

_No response_

### Operating System:

build arm64 ros:humble docker image on amd64 machine with buildkit

### ROS version or commit hash:

ros:humble 8a99e0ca6f90

### RMW implementation (if applicable):

_No response_

### RMW Configuration (if applicable):

_No response_

### Client library (if applicable):

_No response_

### 'ros2 doctor --report' output

<details><summary>ros2 doctor --report</summary>

```console
<COPY OUTPUT HERE>
```
</details>


### Steps to reproduce issue

colcon build

### Expected behavior

succeed to build

### Actual behavior

```
#19 14.28 CMake Error at /opt/ros/humble/share/rcutils/cmake/ament_cmake_export_libraries-extras.cmake:48 (message):
#19 14.28   Package 'rcutils' exports the library 'rcutils' which couldn't be found
#19 14.28 Call Stack (most recent call first):
```

### Additional information

There is no problmen on amd64

I instrumented CMake with strace to see where the behavior diverges. In both cases, the directory listing syscall `getdents64()` on `/opt/ros/humble/lib` returns the same number entries.

In the failure case, there's a failed `mmap()` call around the same time:

```
  mmap(NULL, 36864, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f2832398000
  getdents64(6, 0x7f2832398870 /* 256 entries = [... {d_ino=4543995, d_off=95, d_reclen=40, d_type=DT_REG, d_name="librcutils.so"}...] /*, 32768) = 16032*
  munmap(0x7f2832398000, 36864)           = 0

>> some other allocation happens in the failure case
  mmap(0x7f28379ed000, 143360, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_FIXED_NOREPLACE, -1, 0) = -1 EEXIST (File exists)
  mmap(NULL, 143360, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f283237e000

  mmap(NULL, 36864, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f2832375000
  getdents64(6, 0x7f2832375880 /* 0 entries = [] */, 32768) = 0
  munmap(0x7f2832375000, 36864)           = 0

>> failure case
  write(1, "-- DEBUG: find_library rcutils _"..., 50) = 50
  write(2, "CMake Error at /opt/ros/humble/s"..., 1182) = 1182

>> success case
  close(6)                                = 0
  faccessat(AT_FDCWD, "/opt/ros/humble/lib/librcutils.so", R_OK) = 0
  access("/usr/gnemul/qemu-aarch64/opt/ros/humble/lib/librcutils.so", F_OK) = -1 ENOENT (No such file or directory)
  stat("/opt/ros/humble/lib/librcutils.so", {st_mode=S_IFREG|0644, st_size=93360, ...}) = 0
  write(1, "-- DEBUG: find_library rcutils _"..., 70) = 70
  faccessat(AT_FDCWD, "/opt/ros/humble/lib/librcutils.so", R_OK) = 0   
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -102,8 +102,7 @@ check_library_exists(atomic __atomic_load_8 "" HAVE_LIBATOMICS)
 if(HAVE_LIBATOMICS AND NOT WIN32)
   # Link libatomic into rcutils for fault injection atomic operations.
   # Don't export it - downstream packages get atomic symbols transitively
-  # through librcutils.so, avoiding library resolution issues in containerized
-  # environments. See https://github.com/ros2/rcutils/issues/525
+  # through librcutils.so. See https://github.com/ros2/rcutils/issues/525
   target_link_libraries(${PROJECT_NAME} atomic)
 endif()
 

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -100,8 +100,10 @@ target_link_libraries(${PROJECT_NAME} ${CMAKE_DL_LIBS})
 check_library_exists(atomic __atomic_load_8 "" HAVE_LIBATOMICS)
 
 if(HAVE_LIBATOMICS AND NOT WIN32)
-  # Exporting link flag since it won't pass ament_export_libraries() existence check
-  ament_export_link_flags("-latomic")
+  # Link libatomic into rcutils for fault injection atomic operations.
+  # Don't export it - downstream packages get atomic symbols transitively
+  # through librcutils.so, avoiding library resolution issues in containerized
+  # environments. See https://github.com/ros2/rcutils/issues/525
   target_link_libraries(${PROJECT_NAME} atomic)
 endif()
 

